### ISSUE TITLE
support registering custom type mappings

### ISSUE THREAD

**Issue by ErikGrimes** [dart-lang/polymer_interop#27]
It seems like in order to allow seamless consumption of elements written in JavaScript, there needs to be a way to register unknown JavaScript types so that they are converted to Dart objects.   There's probably a relationship here to custom-element-apigen too.

**Comment by jakemac53** [comment]
cc @jacob314

**Comment by jakemac53** [comment]
Similar to how @CustomElementProxy works but that is just for elements, I assume what you want here is a general purpose thing for any JS Object?

**Comment by ErikGrimes** [comment]
Correct.  If users can register type mappings for any JsObject, then on the Dart side, anything that can comes out of Polymer can be converted to a Dart object instead of JsObject.  If this is done when someone wraps a JavaScript element in custom-element-apigen, then no-one who uses that element should have to know it's a not natively Dart.  I think :)

**Comment by ErikGrimes** [comment]
Also, I'm not sure, but the new js interop may remove the need for this.

**Comment by jacob314** [comment]
Here's an example of what this case should look like like with the new JS interop.
There is no proxying or wrapping. Essentially you define JavaScript interfaces specifying the shape of the JavaScript API you want to access from Dart and then you can access that interface on any JavaScript object passed to Dart where DOM objects are considered JavaScript objects.

```dart
import 'package:js/js.dart';

@Js
class AwesomeCustomElement {
  // all methods must be external.
  external someAwesomeMethod(foo);
  external String get someAwesomeGetter;
} 

void main() {
  AwesomeCustomElement element = querySelector(#yourElementId");
  element.someAwesomeMethod("hello");
}
```
